### ISSUE TITLE
Riivolution builds, but does not work on Console.

### ISSUE THREAD

**Issue by IonicPixels** [AerialX/rawksd#3]
After the updates made, Riivolution builds just fine. I moved the `boot.elf` from the build and put it on my Wii SD Card. I tested on both the OG Wii and Wii U and neither booted Riivolution. What could I have done wrong in the build?

**Comment by IonicPixels** [comment]
here is an issue of the error
![Screenshot 2021-12-04 104705](https://user-images.githubusercontent.com/72478475/144715751-27521203-4ea6-496e-875d-05a98a160e28.png)
:

**Comment by AerialX** [comment]
Heh, just because it compiles doesn't mean it will work! Comparing with the latest official release, does it behave the same or is this message unique to your built version? You might need to set up more logging...

**Comment by IonicPixels** [comment]
this message is unique to the built version. I can give more details through Discord if you would like.

**Comment by AerialX** [comment]
[comment out the define on this line](https://github.com/AerialX/rawksd/blob/master/launcher/source/haxx.cpp#L24) to get more logging out of it. You're mostly on your own though, as I don't actually have a setup for testing anything now.

I wonder if it would be possible to set up more automated testing via dolphin or something... does it even include an ARM emulator for IOS or does it just simulate it with a HAL?

**Comment by IonicPixels** [comment]
I will try that. I'll let you know if that works. I have been testing on console this entire time btw, making sure that the error is not Dolphin only

**Comment by IonicPixels** [comment]
alright I got this:
![Screenshot 2021-12-04 123007](https://user-images.githubusercontent.com/72478475/144718857-a02beaa9-1c53-47cb-94d8-65ad50a60d37.png)

**Comment by MasterofGalaxies** [comment]
I can confirm this issue exists.

> I wonder if it would be possible to set up more automated testing via dolphin or something... does it even include an ARM emulator for IOS or does it just simulate it with a HAL?

Dolphin only simulates IOS.

**Comment by IonicPixels** [comment]
hey @AerialX do you think it would be possible for me to screenshare to you how I am even building Riivolution? I ask so you can tell me what exactly I am doing wrong and if there is anything code-wise that needs to be fixed. If you are open to a screenshare DM me please. (IonicPixels#3139 on Discord)

**Comment by AerialX** [comment]
It doesn't seem like you're doing anything wrong, it's clearly running! If you're really concerned about that you could just [download the elf from github instead](https://github.com/AerialX/rawksd/actions).

> if there is anything code-wise that needs to be fixed

I mentioned fixing warnings earlier, which I've put together some changes for in [extern-syms](https://github.com/AerialX/rawksd/tree/extern-syms). I'm not sure what to blame though since my changes don't seem likely to help (and hopefully I didn't mess up the change and break it even more).
It's just old code that needs to be tested and debugged and I don't have the time or hardware to do that right now, sorry.

> Dolphin only simulates IOS.

Yeah I figured, that's unfortunate.

**Comment by IonicPixels** [comment]
@AerialX do you happen to still have the source files for Riivolution v1.06? iirc it updated the DKP version which could be very helpful for trying to get Riivolution loading on Console

**Comment by IonicPixels** [comment]
I am building the Master branch by the way. I dont know if thats what I am supposed to do or not.

**Comment by AerialX** [comment]
> do you happen to still have the source files for Riivolution v1.06

AFAIK 1.06 was only a build with an updated `libogc` dependency to fix a wiimote crash bug, and did not involve any additional source code changes.

> I am building the Master branch by the way. I dont know if thats what I am supposed to do or not.

That's normally fine, except for the part where I suggested trying a new branch above.

> I just built the program, I can only test on Dolphin at the moment (does not run Riivolution) and I got a black screen.

Yeah there's no point in trying Dolphin really, unless the previous or released elf gets farther which could indicate a difference or problem - but it won't get nearly as far as it did in your above screenshot regardless.

What even is your goal and why are you trying to build it? Despite the recent updates this is all still just a messy old codebase I wrote in high school, it won't be easy to work with if you aren't familiar with C or basic troubleshooting... You might want to start by looking at the [Wiibrew debugging](https://wiibrew.org/wiki/Debugging) page; though it's hacky I've used the [wifi debugger stub](https://wiibrew.org/wiki/User:Qiang0/Debugging#Using_gdb_to_debug_over_the_network) in the past (but it's usually a pain to get working)...

**Comment by IonicPixels** [comment]
@AerialX I am trying to edit Riivolution. It can be pretty buggy so it would be nice to patch out those bugs. It only makes it a little harder to do that if Riivolution wont boot lol. Could you maybe try building this source like you were building an official release and send it? Idk if GitHub allows .elf files to be sent though. If it does I can try running a build of Riivolution that you build yourself

**Comment by IonicPixels** [comment]
I will say that Galaxy Master commented out *most* of `haxx.cpp` and Riivo actually did load, but as soon as the console was supposed to look for a disc it cut to a black screen, and on Wii U it turns off the console

**Comment by AerialX** [comment]
> I am trying to edit Riivolution. It can be pretty buggy so it would be nice to patch out those bugs

This is also a bug and you seem to be completely out of your element. To patch bugs you kind of have to apply critical thinking methods to the problem to know how to approach it and what to do next - and thus far you only seem to be asking others to do things for you? Fixing a bug involves understanding what differs between what the code is attempting to do vs what it's actually doing - which also requires understanding what it's trying to do.
This problem should certainly be fixed, but your behaviour is just confusing as it doesn't match up with what I'd expect from someone who would actually want to work with the code.

That said, feel free to file issues on this repo for known bugs that you plan on fixing, it'd be nice to properly keep track of existing bugs and have others potentially weigh in on them!

> Could you maybe try building this source like you were building an official release and send it?

I [already linked you to where you can download official builds from github](https://github.com/AerialX/rawksd/actions), just click on the relevant commit. This seems unlikely to help though it's worth comparing your version against.

**Comment by IonicPixels** [comment]
@AerialX just so you can understand me better, I'll break the confusion.

>I suggest trying out the extern-syms branch if you haven't had a chance to yet.

I understand how to patch bugs, and I can build it fine, its just the loading that is the issue. I heard somewhere that Riivo v1.06 updated the dkp version used. I have built `extern-syms` and Riivolution still did not load.
 
> and thus far you only seem to be asking others to do things for you?

It may seem that way, but I assure you, its not like that. All I know is that I'd need a working Riivolution build so I could actually start editing Riivo, and maybe even pull request the edits here. As I said, I heard that the dkp version was updated on v1.06 which is why I say it could be very helpful to commit the v1.06 files. If it is not possible to commit the Riivolution v1.06 files, then I guess I could try to look for an ancient version of dkp. I noticed that you changed the version in  `Meta.xml` to be `v1.07` which makes me believe that you want to make an update to Riivo. Knowing the fact that Riivo does not actually work on Console yet makes me a little confused as to how that will work out. I could try commenting way too many things out of haxx.cpp and see if I can get the program to load but not crash, though I'd like to know a better solution than just playing with commenting in that one file. I know that its something in `haxx.cpp` as when that file was mostly commented, Riivolution actually loaded on the Wii and Wii U, but immediately crashed on Wii and powered off the Wii U when the console was supposed to  look for the disk of a game. 

I do have the official v1.06 release installed on both my Wii and vWii Menus and it works just fine. I have also tried using the latest build from the github, and it still gives me the exact same error. I am using the exact same Console throughout my tests and not a single build, except the build from 2013, has worked thus far.

> That said, feel free to file issues on this repo for known bugs that you plan on fixing, it'd be nice to properly keep track of existing bugs and have others potentially weigh in on them!

I do have a fork on my GitHub where I could also just pull request the changes here when I start making changes. It would be nice to have push access to this repo, but I wouldn't be able to commit anything until it all works, even if I did commit things I would make a separate branch.  

>This problem should certainly be fixed, but your behaviour is just confusing as it doesn't match up with what I'd expect from someone who would actually want to work with the code.

I work with code a lot, surprisingly. I always make sure that the program is working as it should before committing, so I know that other people who try to build from source or even make a fork will not have the issue of the program not working as it should. The reason I am asking for a lot of help in this issue is because Riivolution wont load. If the program loaded just fine, this issue probably never would have existed lol. 

I hope I explained myself well in this comment. Sorry for the big wall of text though...

**Comment by MasterofGalaxies** [comment]
@AerialX Do you think you have an unstripped elf of an old working build?

**Comment by AerialX** [comment]
> I understand how to patch bugs, and I can build it fine, its just the loading that is the issue. I heard somewhere that Riivo v1.06 updated the dkp version used. I have built extern-syms and Riivolution still did not load.

Mm, so I have no more obvious suggestions then, sorry. Fixing this issue requires debugging and looking deeper into it.

> I heard that the dkp version was updated on v1.06 which is why I say it could be very helpful to commit the v1.06 files

There are no 1.06 files other than the README/xml changes. That dkp used back then still worked with the original codebase. Keep in mind that 1.06 was released like 8 years ago, which makes it still a rather ancient dkp/libogc.

> I noticed that you changed the version in Meta.xml to be v1.07 which makes me believe that you want to make an update to Riivo

That was more for housekeeping than anything else. 1.06 was finalized years ago, so any test builds from this point onward should not be marked as such. If we do end up landing new bugfixes or improvements, I'll make a release and bump master to 1.08. If not, it could stay unreleased at 1.07 eternally!

> I know that its something in haxx.cpp as when that file was mostly commented, Riivolution actually loaded on the Wii and Wii U, but immediately crashed on Wii and powered off the Wii U when the console was supposed to look for the disk of a game.

Or um, you should know it's something in haxx because the error message tells you that..?

So here's why I say you're not approaching the problem in a way that one would when fixing a bug:
- what haxx does is integral to the entire project: it sets up IOS so that when launching a game, it can patch the kernel and load its own modules that take care of all the important things that riiv does. It removes the security protections that IOS normally provides to the system.
- you can't just remove code you don't understand and expect the rest of the program to work properly? Again I suggest trying to actually understand what it's trying to do, and why it's failing to do it.
    - as you've discovered if you just try to comment it out, Riiv fails to do anything useful because it can't load the code it needs to into IOS
    - if you do want it to work when commented out, you'll need to recreate the conditions it sets up in some other way (or you know, just fix it)
- the screenshot you provided explains part of what is going wrong. We don't know why the values don't match up though; this is the mystery and bug that needs to be solved here

> I do have a fork on my GitHub where I could also just pull request the changes here when I start making changes

Note that pull requests are separate from issues though! The former fix the latter, and generally it's nice to be able to point to a specific issue number that documents the bug that a PR fixes.

in other words: please do tell me what bugs you (and others) have encountered via issues, even if you don't have a fix for them yet!

I'm happy to take PRs either way btw, ideally they're tested of course but they're still helpful as long as they demonstrate and fix some sort of flaw in the logic.

> If the program loaded just fine, this issue probably never would have existed lol.

Yeah that's... what issues are for. I'd certainly like to get this codebase in a working condition and be able to close this issue! Unfortunately we can't right now :(

> Do you think you have an unstripped elf of an old working build?

Unfortunately no, at this point all that's left of the prior builds is what was released in the zip files.

**Comment by IonicPixels** [comment]
> I'd certainly like to get this codebase in a working condition and be able to close this issue!

I'd also like to help. I think if Me, you, Galaxy Master, and a few others work together we can get this repo working!

> you can't just remove code you don't understand and expect the rest of the program to work properly? Again I suggest trying to actually understand what it's trying to do, and why it's failing to do it.

The code that was removed was mainly for a test, to see if it would so anything. it was pretty much just the defines of `babelfish` in Haxx.cpp. 

I previously asked if you could build Riivolution and send it so I could see if it works. I honestly think its better to have a build by you as well as the builds made by me and the build made by GitHub actions so we can know for sure if its a dkp error, or me, a few others, and GitHub actions failing to build properly. If a build by you does not work, then that means it is definitely an error with dkp and not an error with the way we are building. Maybe try building both the regular version of Riivo and the Smash Brawl version of Riivo to see if there is any difference in the way the program reacts (there likely will not be, but I do know that some cpp files are different than the regular version so its worth a shot). Just to be clear: I am not asking you to "build the program because I cant do it" I am asking you to build Riivolution so I can debug it and compare it with my builds and GitHub actions builds.

**Comment by IonicPixels** [comment]
@AerialX another thing: Are you the only person still looking at Riivolutions source code because of this issue? or are the other 3 devs also looking at the code to try to solve what the issue is? I ask because I know its a little over-whelming to look at code like this and try to figure out what the cause of an error like this on your own.

**Comment by AerialX** [comment]
> The code that was removed was mainly for a test, to see if it would so anything. it was pretty much just the defines of babelfish in Haxx.cpp.

Well that doesn't make sense, given that those defines are already omitted from the code? It should not be `define`d, and nothing related to it should be changed except for the part where it removes the `printf` debugging lines. Yes, removing code for the sake of testing is fine; it just doesn't tell you anything more than the `printf`s already do though fyi.

> I think if Me, you, Galaxy Master, and a few others work together we can get this repo working!

I mean I can't because I don't have the time nor a console to test with. It just feels like you need to learn more C before you can tbh, but feel free to prove me wrong! I'd suggest just letting galaxy master ask the questions fwiw, because yeah they might be able to help.

> I'd gladly help maintain it if you'd like. (once its working lol)

Again, I'm happy to take PRs if you provide fixes and improvements but you clearly do not have the experience necessary to maintain a project like this; understanding how it works is kind of a requirement.

> If a build by you does not work, then that means it is definitely an error with dkp and not an error with the way we are building.

Again, you're not approaching the problem in a logical way? I don't have any means of producing a "better quality" build than you or github do. I don't have a devkitPro setup from 2010. I (or someone else) could try to put one together, but I don't feel like there's much value in spending time on that - because ultimately, this problem should be fixed properly instead.

> If a build by you does not work, then that means it is definitely an error with dkp and not an error with the way we are building

It is a problem with the code in combination with the dkp toolchain/compiler. That is what those in the industry call a bug!

> Maybe try building both the regular version of Riivo and the Smash Brawl version of Riivo to see if there is any difference in the way the program reacts (there likely will not be, but I do know that some cpp files are different than the regular version so its worth a shot).

The smash version only changes menus for cosmetic reasons. It does not include any sort of relevant haxx or launcher logic changes AFAIK, so I would not suggest bothering with it.

> Are you the only person still looking at Riivolutions source code because of this issue?

I've not been in contact with anyone in years, but please don't bother more people over this though, it's not like you need someone specific to solve this problem. You just need someone who knows C, it's not that complicated..?

> I am asking you to build Riivolution so I can debug it and compare it with my builds and GitHub actions builds.

The problem is that the questions you ask reveal that you clearly have no idea what to do or what is going on with the program's execution. Which is okay, but like... I can't help you if you don't know how to fix C code, and you're just wasting time by asking for these things.

**Comment by IonicPixels** [comment]
> I've not been in contact with anyone in years, but please don't bother more people over this though, it's not like you need someone specific to solve this problem. You just need someone who knows C, it's not that complicated..?

by that I meant that you and the other devs know the source of Riivo the best, while others like me and Galaxy Master dont know it well, even though we both know C, although Galaxy Master knows C better than I do, he is the one who commented a lot of `haxx.cpp`. I'll send that file if GitHub allows it. 

> I mean I can't because I don't have the time nor a console to test with. It just feels like you need to learn more C before you can tbh, but feel free to prove me wrong! I'd suggest just letting galaxy master ask the questions fwiw, because yeah they might be able to help.

I feel like neither of us would be able to help without either of us knowing the code well. I'd have to look at the `MAKE` files in `rawksd/launcher` to see all of the files called for Riivo, (which should just be all files in the `launcher` folder, but I could be wrong). After separating Riivolution from rawksd entirely then I'll be able to look at all the code for Riivo properly  and know where all the functions are and what they do, which will help me debug easier. I am speaking for both me and Galaxy Master based on both of us building and failing and also even talking about it through Discord. 

> Well that doesn't make sense, given that those defines are already omitted from the code? It should not be defined, and nothing related to it should be changed except for the part where it removes the printf debugging lines. Yes, removing code for the sake of testing is fine; it just doesn't tell you anything more than the printfs already do though fyi.

iirc it was not the whole function, but most of it was commented. I was also provided with a `haxx.patch` file from Galaxy Master, which I will also send if GitHub allows me. 

> Again, I'm happy to take PRs if you provide fixes and improvements but you clearly do not have the experience necessary to maintain a project like this; understanding how it works is kind of a requirement.

As I said, it would be easier for me to understand if Riivo was working.

> Again, you're not approaching the problem in a logical way? I don't have any means of producing a "better quality" build than you or github do. I don't have a devkitPro setup from 2010. I (or someone else) could try to put one together, but I don't feel like there's much value in spending time on that - because ultimately, this problem should be fixed properly instead.

As much as I agree that it should be fixed properly, I never asked for a "better quality" build. just a build in general by you, as out of the 3 of us, (me, you, Galaxy Master) you are the one who knows the code the best.

> The smash version only changes menus for cosmetic reasons. It does not include any sort of relevant haxx or launcher logic changes AFAIK, so I would not suggest bothering with it.

This is good information to know, thanks. I will not bother with it then.

> The problem is that the questions you ask reveal that you clearly have no idea what to do or what is going on with the program's execution. Which is okay, but like... I can't help you if you don't know how to fix C code, and you're just wasting time by asking for these things.

Maybe you could help me understand the code better than I do? If I were to try to fix the code without knowing what critical code functions do could end up in me breaking it more. 

> I can help answer if you have specific technical questions, but other than that the best I can do for you is hope I'll have access to a Wii over the holidays and maybe I can try to look into fixing this then.

If you were to have access to a Wii over the holidays, that would be great! If not, that would be unfortunate, and make solving this issue a little more difficult. 

I am unable to share the code files here on GitHub, please DM me at `IonicPixels#3139` and I will send you the files there.

**Comment by AerialX** [comment]
> by that I meant that you and the other devs know the source of Riivo the best, while others like me and Galaxy Master dont know it well, even though we both know C

If you can't read haxx.cpp, and follow the code+logs+comments (cross-referenced with the logs in your screenshot!) to understand what it's doing... I'm not sure how else to tell you this but: you do not know C.
(do you even know what the error message logs are telling you about what's wrong..?)

> I feel like neither of us would be able to help without either of us knowing the code well

Yes, you need to read the code to learn what it's doing and how it works. If you're unable to do that, then you're far too inexperienced to work with the codebase.

> which should just be all files in the launcher folder, but I could be wrong

They indeed already are separated. Everything under the `rawksd/` folder is rawk-specific, everything under `launcher/` is Riivolution. `rawksd` builds on top of `launcher` and is irrelevant. In fact essentially everything outside of `haxx.cpp` is irrelevant wrt this specific issue.

> I never asked for a "better quality" build. just a build in general by you, as out of the 3 of us, (me, you, Galaxy Master) you are the one who knows the code the best.

You're still missing the point though..? I do not have the ability to make builds differently from you or github. There's nothing special about my setup. If anything, the GitHub builds should be regarded as the ideal environment. The only environment that is different from any of ours would be one from 2010~2013 paired with an older commit of the code.

> I am unable to share the code files here on GitHub,

You could just push them to your fork and link to the commit, but I repeat myself yet again... Showing me the parts you commented out doesn't help either of us.

**Comment by IonicPixels** [comment]
> If you can't read haxx.cpp, and follow the code+logs+comments (cross-referenced with the logs in your screenshot!) to understand what it's doing... I'm not sure how else to tell you this but: you do not know C.
(do you even know what the error message logs are telling you about what's wrong..?)

For me to lean how a program works, I need to see the program working and be able to see my edits, I do have knowledge in C, while it may not be that great, I am still able to read/write C.

> Yes, you need to read the code to learn what it's doing and how it works. If you're unable to do that, then you're far too inexperienced to work with the codebase. 

As I said, I need to be able to see my edits working on a program so I can understand the code better.

> They indeed already are separated. Everything under the rawksd/ folder is rawk-specific, everything under launcher/ is Riivolution. rawksd builds on top of launcher and is irrelevant. In fact essentially everything outside of haxx.cpp is irrelevant wrt this specific issue. After fixing this issue, it's possible that launcher.cpp will need fixing in a similar fashion, and then the *module sub-projects (and the associated libios) after that may need updating too!

What I meant was, isolate Riivolution and its dependencies from rawksd so I'd be able to tell what is *only* Riivolutions code. along with its dependancies, as I have no need to look into rawksd.

> You're still missing the point though..? I do not have the ability to make builds differently from you or github. There's nothing special about my setup. If anything, the GitHub builds should be regarded as the ideal environment. The only environment that is different from any of ours would be one from 2010~2013 paired with an older commit of the code.

Alright then, I will look into finding the real cause of the issue when I have a chance. I'll ask questions if needed.

> You could just push them to your fork and link to the commit, but I repeat myself yet again... Showing me the parts you commented out doesn't help either of us, and in general commenting out code isn't going to help fix the problem here (beyond like... very precise fixes).

Ok, fair enough, I honestly did not think about pushing the changes to my repository but now that you mention it, that is a good idea. I only thought you would like to see the code that was commented because you know the code for Riivolution the best, and thought maybe you would be able to see if anything important was removed from the code.

> Please stop mentioning discord, it is a horrible platform that (imo) needs to be abolished.

Oh..... I had no idea you felt that way about Discord. I'll stop mentioning it then.

**Comment by AerialX** [comment]
> For me to lean how a program works, I need to see the program working and be able to see my edits, I do have knowledge in C, while it may not be that great, I am still able to read/write C.

Understandable, but the code here is somewhat complex. You need to *actually understand* the C source code to work with it, particularly how it interacts with pointers and memory models and such. You're unlikely to get far if it makes no sense to you.

> you would be able to see if anything important was removed from the code

The thing is, almost of the code is important? It's all there for a reason (and if something isn't important, removing it won't help you). In order to fix this issue, you need to (at least partially) understand what those reasons are.

I'm going to hide the irrelevant posts in this issue to preserve the sanity of anyone who comes across this. Feel free to continue in [a discussion topic](https://github.com/AerialX/rawksd/discussions), but I really can't do much for you other than answer specific questions about the code, sorry.

**Comment by Lami-Lemmy** [comment]
> The smash version only changes menus for cosmetic reasons. It does not include any sort of relevant haxx or launcher logic changes AFAIK, so I would not suggest bothering with it.

Imma just pop in for a second and mention that the brawl build available on the riivolution archive produces the same error for me as the one in the first screenshot.

**Comment by AerialX** [comment]
> Imma just pop in for a second and mention that the brawl build available on the riivolution archive produces the same error for me as the one in the first screenshot.

Interesting, thanks for the note. IIRC it should be equivalent to Riiv 1.02, though I might've expected you to see a "downgrade" message instead of a "reinstall"? The requirement for a specific IOS37 version does make things a bit tricky, but it doesn't tell you all that much without the extra logging enabled. It could be worth testing with 1.05/1.04 and manually checking/installing IOS37 versions (don't do this on a Wii U) - but really we still need to do a little in-depth debugging.

**Comment by Lami-Lemmy** [comment]
>  though I might've expected you to see a "downgrade" message instead of a "reinstall"?

After testing it again, you're right. Was basing off of memory of a few years ago when I had tried to install it.

**Comment by Lami-Lemmy** [comment]
After having commented out the define line in haxx, the log was not the same. Weirdly it does not give the message saying that AHBPROT was already disabled, even though I was launching from HBC
![PXL_20211217_225117023.jpg](https://user-images.githubusercontent.com/58743698/146617005-7457beaf-62bb-454f-8357-2fdb00087ba4.jpg)

**Comment by MasterofGalaxies** [comment]
I compiled Riivolution with devkitPro and libraries from when v1.06 was released. It doesn't work entirely; the console still powered off shortly after the GUI appeared (just like in builds with modern DKP and libs, as was reported here earlier).

However, the haxx.cpp issues did not occur. To fix those, I'd suggest bindiffing haxx.o from my build with haxx.o as built with modern DKP (as I was going to do myself, however I have not had time to work on this lately).

[haxx.o.zip](https://github.com/AerialX/rawksd/files/7738501/haxx.o.zip)

**Comment by IonicPixels** [comment]
When I changed the errors to return as if they were working properly, I was able to get Riivo to reboot the console instead of power off. I was using modern devkitPro. I then set all returns back to `-1` and added some prints to where the errors occur, which happens to be at `do_exploit` (where it breaks, and forces the exit to HBC)

**Comment by Fynn93** [comment]
Does anyone know how to fix this?

**Comment by AerialX** [comment]
I do have a Wii to test with now, will probably get to it within a few weeks fwiw.

**Comment by Lami-Lemmy** [comment]
I found [this repository](https://github.com/zurgeg/riivolution) which has some work done, notably to the problematic haxx.cpp. I haven't tried to compile it yet so I don't know if it was successful but it might be.

**Comment by Exorcism0666** [comment]
Note: vWii / PAL

Boot from the HBL vWii

![IMG20220130202914.jpg](https://user-images.githubusercontent.com/46537034/151714452-c7aad1ad-db48-4b2f-b2ff-a69221081b80.jpg)

**Comment by MasterofGalaxies** [comment]
The system shutting down or rebooting, as I and others have mentioned here, is caused by CheckShutdown() at

https://github.com/AerialX/rawksd/blob/14b6f27cfb3ab1e4a2288a34f145f844ca96bd4a/launcher/source/init.cpp#L74

This function is presumably supposed to shutdown the system when the power button has been pressed, but for some reason instead shuts down the system every time it's called. Didn't look into that; just disabled that function.

After doing that, I compiled Riivolution with old versions of devkitPPC and the libraries it uses, and got a build that was able to launch a game with patches on a USB drive. (It did not work with an SD card.)

**Comment by Fynn93** [comment]
So I replaced haxx.cpp with this one [haxx.cpp](https://github.com/zurgeg/riivolution/blob/master/launcher/source/haxx.cpp) It compiled successfully, but it still doesnt work.

**Comment by AerialX** [comment]
So there have been a bunch of distinct issues being reported, I'm kind of tracking them here:
- [x] [CheckShutdown](https://github.com/AerialX/rawksd/issues/3#issuecomment-1025215229): e347b360ff96950a0886605c3dfb475d05650b7e
- [x] [IOS 37 TMD failed sig check](https://github.com/AerialX/rawksd/issues/3#issuecomment-997077311): 943cb091e8bec97823352f1dc4725249ffeeabac
- [ ] [do_patch 0 failed](https://github.com/AerialX/rawksd/issues/3#issuecomment-986063197)

I encountered the issue reported by Lami-Lemmy, so I fixed that first. I can at least confirm that builds made with modern dkP work well enough now; I launched [Newer](https://newerteam.com/wii/) as a test via SD card and it seemed alright. RiiFS seemed to work as expected too.

My assumption is that the remaining error is unique to iOS37 v5663 - I'm currently testing on a slightly outdated Wii with v5662 but I'll try updating it soon ([dop-mii](https://wiibrew.org/wiki/DOP-Mii:_WiiBrew_Edition) seems like a convenient way to upgrade/downgrade for testing, but **do not** use it on a vWii, heed the warnings, etc).

Have we confirmed that the `do_patch 0 failed` message also appears on the Wii U / vWii when you comment out the printf for the extra debugging messages? I also haven't tested installing the Riiv channel yet, it would be nice to know if that works now?

**Comment by Fynn93** [comment]
So. after I compiled Riivolution and ran it, Riivolution got stuck in and endless loop of "Loading...".

That is another bug (maybe?)

**Comment by AerialX** [comment]
> So, after I compiled Riivolution and ran it, Riivolution got stuck in an endless loop of "Loading...".

Try downloading from [the CI build](https://github.com/AerialX/rawksd/actions/runs/2098795087), and test with different combinations of SD card, no SD card, DVD, no DVD, no USB, etc. That riivolution.zip's boot.elf launches properly like my local builds here. I've only tested SD and WiFi - not sure if USB works but I'd expect it to.

Also what hardware version of Wii, system menu version, etc. are you on?

**Comment by IonicPixels** [comment]
> So there have been a bunch of distinct issues being reported, I'm kind of tracking them here:
> 
> * [x]  [CheckShutdown](https://github.com/AerialX/rawksd/issues/3#issuecomment-1025215229): [e347b36](https://github.com/AerialX/rawksd/commit/e347b360ff96950a0886605c3dfb475d05650b7e)
> * [x]  [IOS 37 TMD failed sig check](https://github.com/AerialX/rawksd/issues/3#issuecomment-997077311): [943cb09](https://github.com/AerialX/rawksd/commit/943cb091e8bec97823352f1dc4725249ffeeabac)
> * [ ]  [do_patch 0 failed](https://github.com/AerialX/rawksd/issues/3#issuecomment-986063197)

what I am getting from this is that do_patch is the only thing preventing Riivo from working perfectly. If I understand this right, is `do_patch` failing sometimes, most of the time, or randomly?

**Comment by AerialX** [comment]
Sorry I have no idea, I never saw it on the Wii I have here so can't reproduce it at all. It might be worth laying out:
- Wii hardware variant/region/version/etc
- Version of IOS37 currently installed on the system
- Method of launching Riiv (HBC? version? SD? USB? Wiiload?)
- Do you have bootmii or some other pervasive mods installed? Anything strange about your system menu or have some sort of cIOS or something?
- Get full logs instead of letting the IOS37 reinstall clobber half the screen

Also maybe try modifying `do_patch` by [changing the check to `if (true) {`](https://github.com/AerialX/rawksd/blob/4ef79e2e234d7c2f19d030c2d9440852b5c73b7a/launcher/source/haxx.cpp#L474) and see if it works anyway without the safety check?

**Comment by IonicPixels** [comment]
oh that is a good idea, I'll mention now my Wii is the first revision (RVL 001), version 4.3U, region being US. IOS37 is currently installed, (Riivolution 1.6 works fine) the method of loading the CI builds and my own builds with HBC, on an SD. I do have Bootmii and Priiloader on my Wii. I'll change the check to `if (true) {` and test the CI build on my Wii when I get home from school.  I'll report back any new information as soon as I can

**Comment by lumentae** [comment]
The CI Build works for me, but self-compiling still doesn't work.

**Comment by lilousky** [comment]
> The CI Build works for me, but self-compiling still doesn't work.

That's most likely an issue in your setup, what's your OS/revision of dkp?

**Comment by lumentae** [comment]
> That's most likely an issue in your setup, what's your OS/revision of dkp?

OS: Manjaro Linux 21.2.5 (based on Arch Linux)
DKP: devkitPPC r40
     devkitARM r57

Also installing the Channel works.

**Comment by IonicPixels** [comment]
> The CI Build works for me, but self-compiling still doesn't work.

I can confirm this, its only self-compiling that fails.
